File: src/test/scala/com/carmanconsulting/akka/scala/PropsTest.scala
package com.carmanconsulting.akka.scala

import org.junit.Test
import akka.actor.Props

class PropsTest extends AkkaTestCase {
  @Test
  def testWithNoArguments(): Unit = {
    val hello = system.actorOf(Props[ParameterizedHelloAkka], "hello")
    hello ! "Akka"
    expectMsg("Hello, Akka!")
  }

  @Test
  def testWithArguments(): Unit = {
    val hello = system.actorOf(Props(classOf[ParameterizedHelloAkka], "Hola, %s!"), "hello")
    hello ! "Akka"
    expectMsg("Hola, Akka!")
  }

  @Test
  def testWithCreator(): Unit = {
    val hello = system.actorOf(Props(PropsTest.create()), "hello")
    hello ! "Akka"
    expectMsg("Bonjour, Akka!")
  }
}

object PropsTest {
  def create() = {
    new ParameterizedHelloAkka("Bonjour, %s!")
  }
}

File: java/src/test/java/com/carmanconsulting/akka/java/PropsTest.java
package com.carmanconsulting.akka.java;

import akka.actor.ActorRef;
import akka.actor.Props;
import akka.japi.Creator;
import org.junit.Test;

public class PropsTest extends AkkaTestCase {

    @Test
    public void testWithNoArguments() {
        ActorRef hello = system().actorOf(Props.create(ParameterizedHelloAkka.class), "hello");
        hello.tell("Akka", testActor());
        expectMsg("Hello, Akka!");
    }

    @Test
    public void testWithArguments() {
        ActorRef hello = system().actorOf(Props.create(ParameterizedHelloAkka.class, "Hola, %s!"), "hello");
        hello.tell("Akka", testActor());
        expectMsg("Hola, Akka!");
    }

    @Test
    public void testWithCreator() {
        ActorRef hello = system().actorOf(Props.create(new HelloAkkaActorCreator()), "hello");
        hello.tell("Akka", testActor());
        expectMsg("Bonjour, Akka!");
    }

    public static class HelloAkkaActorCreator implements Creator<ParameterizedHelloAkka> {
        @Override
        public ParameterizedHelloAkka create() throws Exception {
            return new ParameterizedHelloAkka("Bonjour, %s!");
        }
    }
}
